modules/apps/callbacks/v2/ibc_middleware.go

```diff
@@ -0,0 +1,403 @@
+	maxCallbackGas uint64
+}
+
+// NewIBCMiddleware creates a new IBCMiddleware instance given the keeper and underlying application.
+// The underlying application must implement the required callback interfaces.
+func NewIBCMiddleware(
+	app api.IBCModule, writeAckWrapper api.WriteAcknowledgementWrapper,
+	contractKeeper types.ContractKeeper, chanKeeperV2 types.ChannelKeeperV2, maxCallbackGas uint64,
+) IBCMiddleware {
+	packetDataUnmarshalerApp, ok := app.(types.CallbacksCompatibleModuleV2)
+	if !ok {
+		panic(fmt.Errorf("underlying application does not implement %T", (*types.CallbacksCompatibleModule)(nil)))
+	}
+
+	if contractKeeper == nil {
+		panic(errors.New("contract keeper cannot be nil"))
+	}
+
+	if maxCallbackGas == 0 {
+		panic(errors.New("maxCallbackGas cannot be zero"))
+	}
+
+	return IBCMiddleware{
+		app:             packetDataUnmarshalerApp,
+		writeAckWrapper: writeAckWrapper,
+		contractKeeper:  contractKeeper,
+		chanKeeperV2:    chanKeeperV2,
+		maxCallbackGas:  maxCallbackGas,
+	}
+}
+
+// WithWriteAckWrapper sets the WriteAcknowledgementWrapper for the middleware.
+func (im *IBCMiddleware) WithWriteAckWrapper(writeAckWrapper api.WriteAcknowledgementWrapper) {
+	im.writeAckWrapper = writeAckWrapper
+}
+
+// GetWriteAckWrapper returns the WriteAckWrapper
+func (im *IBCMiddleware) GetWriteAckWrapper() api.WriteAcknowledgementWrapper {
+	return im.writeAckWrapper
+}
+
+// OnSendPacket implements source callbacks for sending packets.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback returns an error, panics, or runs out of gas, then
+// the packet send is rejected.
+func (im IBCMiddleware) OnSendPacket(
+	ctx context.Context,
+	sourceClient string,
+	destinationClient string,
+	sequence uint64,
+	payload channeltypesv2.Payload,
+	signer sdk.AccAddress,
+) error {
+	err := im.app.OnSendPacket(ctx, sourceClient, destinationClient, sequence, payload, signer)
+	if err != nil {
+		return err
+	}
+
+	packetData, err := im.app.UnmarshalPacketData(payload)
+	// OnSendPacket is not blocked if the packet does not opt-in to callbacks
+	if err != nil {
+		return nil
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	cbData, err := types.GetCallbackData(
+		packetData, payload.GetVersion(), payload.GetSourcePort(),
+		sdkCtx.GasMeter().GasRemaining(), im.maxCallbackGas, types.SourceCallbackKey,
+	)
+	// OnSendPacket is not blocked if the packet does not opt-in to callbacks
+	if err != nil {
+		return nil
+	}
+
+	callbackExecutor := func(cachedCtx sdk.Context) error {
+		return im.contractKeeper.IBCSendPacketCallback(
+			cachedCtx, payload.SourcePort, sourceClient, clienttypes.Height{}, 0, payload.Value, cbData.CallbackAddress, cbData.SenderAddress, payload.Version,
+		)
+	}
+
+	err = types.ProcessCallback(sdkCtx, types.CallbackTypeSendPacket, cbData, callbackExecutor)
+	// contract keeper is allowed to reject the packet send.
```

Comment:
contract keeper involved here? Not sure I understand

Reply:
contractKeeper is wrapped in callback executor